ipn_main_handler.php

```diff
@@ -1,517 +1,584 @@
+            ipn_debug_email('IPN NOTICE :: Storing order/update details for order #' . $ordersID . ' txn_id: ' . $_POST['txn_id'] . ' PP IPN ID: ' . $paypalipnID);
+            if ($txn_type === 'parent') {
+                $sql_data_array = ipn_create_order_array($ordersID, $txn_type);
+                zen_db_perform(TABLE_PAYPAL, $sql_data_array);
+                $paypalipnID = $db->Insert_ID();
+            } else {
+                $sql_data_array = ipn_create_order_update_array($txn_type);
+                zen_db_perform(TABLE_PAYPAL, $sql_data_array, 'update', "txn_id='" . ($txn_type === 'cleared-authorization' ? $_POST['parent_txn_id'] : $_POST['txn_id']) . "'");
+                $sql = "SELECT paypal_ipn_id FROM " . TABLE_PAYPAL . " WHERE txn_id=:txn:";
+                $sql = $db->bindVars($sql, ':txn:', $_POST['txn_id'], 'string');
+                $result = $db->Execute($sql, 1);
+                $paypalipnID = $result->fields['paypal_ipn_id'];
+            }
+            $sql_data_array = ipn_create_order_history_array($paypalipnID);
+            zen_db_perform(TABLE_PAYPAL_PAYMENT_STATUS_HISTORY, $sql_data_array);
+            ipn_debug_email(
+                'IPN NOTICE :: Added PP status-history record for order #' . $ordersID . ' txn_id: ' . $_POST['txn_id'] . ' (updated/child) PP IPN ID: ' . $paypalipnID
+            );
+
+            switch ($txn_type) {
+                case 'voided':
+                case ($_POST['payment_status'] === 'Refunded' || $_POST['payment_status'] === 'Reversed' || $_POST['payment_status'] === 'Voided'):
+                    //payment_status=Refunded or payment_status=Voided
+                    $new_status = defined('MODULE_PAYMENT_PAYPALWPP_REFUNDED_STATUS_ID') ? MODULE_PAYMENT_PAYPALWPP_REFUNDED_STATUS_ID : 1;
+                    if (defined('MODULE_PAYMENT_PAYPAL_REFUND_ORDER_STATUS_ID') && (int)MODULE_PAYMENT_PAYPAL_REFUND_ORDER_STATUS_ID > 0 && !$isECtransaction) {
+                        $new_status = MODULE_PAYMENT_PAYPAL_REFUND_ORDER_STATUS_ID;
+                    }
+                    break;
+                case 'echeck-denied':
+                case 'denied-echeck':
+                case 'failed-echeck':
+                    //payment_status=Denied or failed
+                    $new_status = ($isECtransaction ? MODULE_PAYMENT_PAYPALWPP_REFUNDED_STATUS_ID : MODULE_PAYMENT_PAYPAL_REFUND_ORDER_STATUS_ID);
+                    break;
+                case 'echeck-cleared':
+                    $new_status = (int)MODULE_PAYMENT_PAYPAL_ORDER_STATUS_ID;
+                    break;
+                case ($txn_type === 'express-checkout-cleared' || str_starts_with($txn_type, 'cleared-')):
+                    //express-checkout-cleared
+                    $new_status = ($isECtransaction && defined('MODULE_PAYMENT_PAYPALWPP_ORDER_STATUS_ID') ? MODULE_PAYMENT_PAYPALWPP_ORDER_STATUS_ID : MODULE_PAYMENT_PAYPAL_ORDER_STATUS_ID);
+                    if (empty($new_status)) {
```

Comment:
Should `MODULE_PAYMENT_PAYPAL_ORDER_STATUS_ID` in above statement be casted to int?

Reply:
IMO, that entire conditional set into `$new_status` should be cast to int, not just that one constant.

Reply:
Okay, will do that for consistency and for clarity to the human reader.

Everywhere the value gets used already first casts it to `int`. eg:
https://github.com/zencart/zencart/blob/445529df14235b84547b5c35f3b2d8341380c859/includes/functions/functions_osh_update.php#L85

Reply:
Done.